entity-server-api/src/main/java/org/terracotta/entity/ServiceProvider.java

```diff
@@ -22,46 +22,69 @@
+ *  implementation which has shared state must ensure that it can manage this state in a thread safe way (subject to whatever
+ *  thread safety its own state requires).
  */
 public interface ServiceProvider {
+  /**
+   * PLATFORM_CONSUMER_ID is the consumerID used by the platform, itself, when it wants to interact with services.  Normal
+   *  entity instances use positive consumeID values.
+   */
+  public final long PLATFORM_CONSUMER_ID = 0; 
 
   /**
-   * The platform configuration based on which the Service provider can choose to initialize itself.
-   *
-   * @param configuration platform configuration
-   * @return true if provider has successfully been initialized with the provided configuration
+   * The platform configuration based on which the receiver can choose to initialize itself.
+   * This is called once on each provider instance, at server start-up time, before any other methods have been called on the
+   *  instance.
+   * Note that it is possible for there to be multiple instances of the same ServiceProvider if the server had multiple
+   *  configuration stanzas handled by the same ServiceProvider class.  As the entities requesting services do not have direct
+   *  access to the set of initialized ServiceProviders, they do not require special accounting for this possibility.
+   * 
+   * @param configuration The configuration for this ServiceProvider instance.
+   * @return True if provider has successfully been initialized with the provided configuration, false if this instance
+   *  shouldn't be registered for the given configuration (it will be discarded).
    */
   boolean initialize(ServiceProviderConfiguration configuration);
 
   /**
-   * Get an instance of service from the provider.
-   *
-   * @param consumerID The unique ID used to name-space the returned service
-   * @param configuration Service configuration which is to be used
-   * @return service instance
+   * Get an instance of a service from the provider.
+   * Note that this method can be called any number of times, over the life of a ServiceProvider instance, potentially
+   *  concurrently or multiple times for the same consumerID.  If the implementation wishes to intern, or otherwise track,
+   *  the instances it returns, it will need to ensure its approach is robust against concurrent calls.
+   * Note that the consumerID is a unique number for each entity instance which is used to effectively name-space the service
+   *  instances under the entities.  An entity instance will have the same consumerID for its entire life (including after
+   *  restart or fail-over).  The value is always non-negative.
+   * 
+   * @param consumerID The unique, non-negative, ID used to name-space the returned service for a specific consuming entity
+   *  (or the platform).
+   * @param configuration Service configuration which is to be used.
+   * @return The service instance (which may be either new or a cached/interned instance for this consumerID).
```

Comment:
What do you return if a specific instance cannot provide the service and another instance should be providing it instead?